Source/WebCore/Modules/gamepad/GamepadManager.cpp

```diff
@@ -115,30 +116,32 @@ void GamepadManager::platformGamepadDisconnected(PlatformGamepad& platformGamepa
 
         // This LocalDOMWindow's Navigator might not be accessible. e.g. The LocalDOMWindow might be in the back/forward cache.
         // If this happens the LocalDOMWindow will not get this gamepaddisconnected event.
-        NavigatorGamepad* navigator = navigatorGamepadFromDOMWindow(*window);
-        if (!navigator)
-            continue;
+        Ref navigator = window->navigator();
 
         // If this Navigator hasn't seen gamepads yet then its Window should not get the disconnect event.
-        if (m_gamepadBlindNavigators.contains(*navigator))
+        if (m_gamepadBlindNavigators.contains(navigator.get()))
             continue;
 #if PLATFORM(VISION)
-        if (m_gamepadQuarantinedNavigators.contains(*navigator))
+        if (m_gamepadQuarantinedNavigators.contains(navigator.get()))
             continue;
 #endif
 
-        Ref<Gamepad> gamepad(navigator->gamepadFromPlatformGamepad(platformGamepad));
+        auto* navigatorGamepad = NavigatorGamepad::from(navigator);
+        if (!navigatorGamepad)
+            continue;
+
+        Ref gamepad = navigatorGamepad->gamepadFromPlatformGamepad(platformGamepad);
 
-        navigator->gamepadDisconnected(platformGamepad);
-        notifiedNavigators.add(*navigator);
+        navigatorGamepad->gamepadDisconnected(platformGamepad);
+        notifiedNavigators.add(navigator.get());
 
         window->dispatchEvent(GamepadEvent::create(eventNames().gamepaddisconnectedEvent, gamepad.get()), window->document());
     }
 
     // Notify all the Navigators that haven't already been notified.
-    for (auto& navigator : m_navigators) {
-        if (!notifiedNavigators.contains(navigator))
-            navigator.gamepadDisconnected(platformGamepad);
+    for (Ref navigator : m_navigators) {
+        if (!notifiedNavigators.contains(navigator.get()))
+            NavigatorGamepad::from(navigator)->gamepadDisconnected(platformGamepad);
```

Comment:
Is it really guaranteed that each `Navigator` here would have `NavigatorGamepad`? Wouldn't it be more conservative to `nullptr` check here?

Reply:
oh I see. Yeah, making the function return a reference would make it more obvious to the reader that it can't be `nullptr`.